src/repart/repart.c

```diff
@@ -5069,17 +5088,26 @@ static int partition_format_verity_sig(Context *context, Partition *p) {
 
         assert_se((whole_fd = fdisk_get_devfd(context->fdisk_context)) >= 0);
 
-        r = sign_verity_roothash(&hp->roothash, context->certificate, context->private_key, &sig);
-        if (r < 0)
-                return r;
+        if (arg_verity_settings.root_hash_size == 0) {
+                r = sign_verity_roothash(&hp->roothash, context->certificate, context->private_key, &sig_free);
+                if (r < 0)
+                        return r;
+                sig = sig_free;
+                roothash = hp->roothash;
+        } else {
+                sig.iov_base = arg_verity_settings.root_hash_sig;
+                sig.iov_len = arg_verity_settings.root_hash_sig_size;
+                roothash.iov_base = arg_verity_settings.root_hash;
+                roothash.iov_len = arg_verity_settings.root_hash_size;
```

Comment:
hmm, i think it would be a lot nicer if we'd parse the signature, and then search for the root hash we need in it, and complain if we can't find it.

i.e. get rid of the --root-hash= parameter, just keep the --root-hash-sig= paramater, and match up root hashes properly?

Reply:
mmh if the roothash was available, wouldn't have been in `hp->roothash`? I needed to add the parameter as that was unset.

Also without the roothash how do I match it? The signature is detached, it does not have data?

Reply:
Yeah as far as I can tell it's not. The signature is pkcs7 detached, so it only contains the signer cert + rsa signature blob.